File: src/utils/fileUtils.ts
import fs from 'node:fs';
import path from 'node:path';

import Handlebars from 'handlebars';

export interface ReviewMetadata {
  timestamp: string;
  llmProvider: string;
  llmModel: string;
  mrMode: string;
  gitPlatform: string;
  commandLine: string;
  [key: string]: unknown;
}

interface ReviewData {
  metadata: ReviewMetadata;
  feedback: object;
}

/**
 * File utility class for file operations
 */
export class FileUtils {
  /**
   * Ensures a directory exists
   * @param dirPath - Directory path to ensure
   * @returns True if directory was created, false if it already existed
   * @throws If the directory cannot be created
   */
  private ensureDirectoryExists(dirPath: string): boolean {
    if (fs.existsSync(dirPath)) {
      return false;
    }

    fs.mkdirSync(dirPath, { recursive: true });
    return true;
  }

  /**
   * Generate a timestamped filename for a review
   * @param reviewsDir - Base directory for reviews
   * @returns Full path with timestamped filename
   */
  private generateReviewFilename(reviewsDir: string): string {
    const timestamp = new Date().toISOString().replace(/:/g, '-');
    return path.join(reviewsDir, `review-${timestamp}.json`);
  }

  /**
   * Parse feedback string to object if needed
   * @param feedback - Feedback as string or object
   * @returns Parsed feedback as object
   * @throws If feedback string cannot be parsed as JSON
   */
  private parseFeedback(feedback: string | object): object {
    if (typeof feedback !== 'string') {
      return feedback;
    }

    try {
      return JSON.parse(feedback);
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : 'Unknown error';
      
      // Create a fallback structured response when JSON parsing fails
      const fallbackResponse = {
        overview: {
          summary: "Code review completed but response was not in expected JSON format",
          riskLevel: "medium",
          recommendedAction: "comment"
        },
        fileReviews: [],
        testReview: {
          compliance: "unknown",
          missingTests: ["Unable to parse test analysis from response"],
          testQualityIssues: []
        },
        generalFeedback: {
          strengths: [],
          concerns: ["LLM response was not in valid JSON format"],
          suggestions: [
            "Review the raw response manually",
            "Consider adjusting the prompt to enforce JSON output"
          ]
        },
        rawResponse: feedback.substring(0, 1000) // Include first 1000 chars of raw response
      };
      
      console.warn(`Warning: Could not parse LLM response as JSON: ${errorMessage}`);
      console.warn(`Raw response (first 200 chars): ${feedback.substring(0, 200)}...`);
      
      return fallbackResponse;
    }
  }

  /**
   * Create review data object from metadata and feedback
   * @param metadata - Review metadata
   * @param parsedFeedback - Parsed feedback object
   * @returns Combined review data object
   */
  private createReviewData(metadata: ReviewMetadata, parsedFeedback: object): ReviewData {
    return {
      metadata,
      feedback: parsedFeedback,
    };
  }

  /**
   * Write review data to file
   * @param filePath - Path to write the file
   * @param reviewData - Review data to write
   * @throws If the file cannot be written
   */
  private writeReviewFile(filePath: string, reviewData: ReviewData): void {
    fs.writeFileSync(filePath, JSON.stringify(reviewData, null, 2));
  }

  /**
   * Save review data to a JSON file
   * @param reviewsDir - Directory to save reviews
   * @param feedback - The feedback to save
   * @param metadata - Metadata about the review
   * @returns The path where the review was saved
   * @throws If the review dir cannot be created or the file cannot be written
   */
  saveReviewToFile(
    reviewsDir: string,
    feedback: string | object,
    metadata: ReviewMetadata
  ): string {
    try {
      this.ensureDirectoryExists(reviewsDir);

      const filePath = this.generateReviewFilename(reviewsDir);
      const parsedFeedback = this.parseFeedback(feedback);
      const reviewData = this.createReviewData(metadata, parsedFeedback);

      this.writeReviewFile(filePath, reviewData);

      console.log(`Review saved to ${filePath}`);
      return filePath;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : 'Unknown error';
      console.error('Error saving review to file:', errorMessage);
      throw error;
    }
  }

  /**
   * Get all markdown files from a directory
   * @param promptDir - Directory to search for markdown files
   * @returns Sorted array of markdown filenames
   * @throws If the directory cannot be read
   */
  private getMarkdownFiles(promptDir: string): string[] {
    return fs
      .readdirSync(promptDir)
      .filter(file => file.endsWith('.md'))
      .sort(); // Sort to ensure consistent order
  }

  /**
   * Read file content and return as string
   * @param filePath - Path to file
   * @returns File content as string
   * @throws If file cannot be read
   */
  private readFileContent(filePath: string): string {
    return fs.readFileSync(filePath, 'utf8');
  }

  /**
   * Combine multiple file contents into a single string
   * @param promptDir - Directory containing files
   * @param files - Array of filenames to combine
   * @returns Combined content
   */
  private combineFileContents(promptDir: string, files: string[]): string {
    let fullPrompt = '';

    for (const file of files) {
      const filePath = path.join(promptDir, file);
      const content = this.readFileContent(filePath);
      fullPrompt += `${content}\n\n`;
    }

    return fullPrompt.trim();
  }

  /**
   * Validate prompt directory exists
   * @param promptDir - Directory path to validate
   * @throws If directory doesn't exist
   */
  private validatePromptDirectory(promptDir: string): void {
    if (!fs.existsSync(promptDir)) {
      throw new Error(`Prompt directory not found: ${promptDir}`);
    }
  }

  /**
   * Load a directory of prompt files and combine them into a full prompt
   * @param promptDir - Directory containing the prompt files in markdown format
   * @returns The combined prompt template
   */
  loadPromptFromDirectory(promptDir: string): string {
    try {
      this.validatePromptDirectory(promptDir);

      const files = this.getMarkdownFiles(promptDir);

      if (files.length === 0) {
        return this.getBestPracticesPrompt();
      }

      return this.combineFileContents(promptDir, files);
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : 'Unknown error';
      console.error(`Error loading prompt template from directory ${promptDir}:`, errorMessage);
      throw error;
    }
  }

  /**
   * Validate formatter template file exists
   * @param templatePath - Path to validate
   * @throws If file doesn't exist
   */
  private validateTemplateExists(templatePath: string): void {
    if (!fs.existsSync(templatePath)) {
      throw new Error(`Formatter template not found: ${templatePath}`);
    }
  }

  /**
   * Compile Handlebars template from content
   * @param content - Template content
   * @returns Compiled Handlebars template
   */
  private compileTemplate(content: string): Handlebars.TemplateDelegate {
    return Handlebars.compile(content);
  }

  /**
   * Load a formatter template from file
   * @param templatePath - Path to the formatter template file
   * @returns The formatter template
   */
  loadFormatterTemplate(templatePath: string): Handlebars.TemplateDelegate {
    try {
      this.validateTemplateExists(templatePath);

      const content = this.readFileContent(templatePath);
      return this.compileTemplate(content);
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : 'Unknown error';
      console.error(`Error loading formatter template from ${templatePath}:`, errorMessage);
      throw error;
    }
  }

  /**
   * Get a fallback prompt if no files are found
   * @returns A default best practices prompt
   */
  private getBestPracticesPrompt(): string {
    return `Review this code pull request and provide a detailed analysis focusing on:
1. Logical errors and bugs
2. Security vulnerabilities
3. Performance issues
4. Maintainability concerns

Please evaluate according to best practices for the language and framework in use.
Provide specific feedback with line numbers when possible.

Structure your response in JSON format for easier parsing.

CODE DIFF:
{{diff}}`;
  }
}

File: python/tiw/utils/file_utils.py
"""File operations utilities."""

import datetime
import json
import os
import pathlib
from typing import Any, Dict, Optional

from .logging import Logger


class FileUtils:
    """Utility class for file operations."""
    
    def __init__(self, logger: Optional[Logger] = None):
        """Initialize FileUtils.
        
        Args:
            logger: Optional logger instance
        """
        self.logger = logger or Logger()
    
    def load_prompt_from_directory(self, prompt_dir: str) -> str:
        """Load the prompt template from a directory.
        
        Args:
            prompt_dir: Directory containing prompt files
            
        Returns:
            Combined prompt template
            
        Raises:
            FileNotFoundError: If prompt directory doesn't exist
        """
        prompt_dir_path = pathlib.Path(prompt_dir)
        
        if not prompt_dir_path.exists() or not prompt_dir_path.is_dir():
            self.logger.error(f"Prompt directory not found: {prompt_dir}")
            raise FileNotFoundError(f"Prompt directory not found: {prompt_dir}")
        
        # List of prompt sections to load and combine
        sections = [
            "introduction.md",
            "criteria.md",
            "priorities.md",
            "diff.md",
            "output_format.md"
        ]
        
        combined_prompt = ""
        for section in sections:
            section_path = prompt_dir_path / section
            if section_path.exists():
                section_content = section_path.read_text(encoding="utf-8")
                combined_prompt += f"{section_content}\n\n"
                self.logger.debug(f"Loaded prompt section: {section}")
            else:
                self.logger.warn(f"Prompt section not found: {section}")
        
        if not combined_prompt:
            self.logger.error("No prompt sections found")
            raise FileNotFoundError("No prompt sections found in directory")
        
        return combined_prompt
    
    def save_review_to_file(
        self, 
        reviews_dir: str,
        feedback: str,
        metadata: Dict[str, Any]
    ) -> str:
        """Save review to a file.
        
        Args:
            reviews_dir: Directory to save the review
            feedback: LLM review feedback
            metadata: Review metadata
            
        Returns:
            Path to the saved file
        """
        # Create reviews directory if it doesn't exist
        reviews_path = pathlib.Path(reviews_dir)
        reviews_path.mkdir(parents=True, exist_ok=True)
        
        # Generate filename with timestamp
        timestamp = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
        filename = f"review_{timestamp}.json"
        file_path = reviews_path / filename
        
        # Create review object with metadata and feedback
        review_data = {
            "metadata": metadata,
            "feedback": feedback
        }
        
        # Write to file
        with open(file_path, "w", encoding="utf-8") as f:
            json.dump(review_data, f, indent=2)
        
        self.logger.info(f"Review saved to: {file_path}")
        return str(file_path)
    
    def ensure_directory_exists(self, directory_path: str) -> str:
        """Ensure that a directory exists, creating it if necessary.
        
        Args:
            directory_path: Path to the directory
            
        Returns:
            Absolute path to the directory
        """
        path = pathlib.Path(directory_path).resolve()
        path.mkdir(parents=True, exist_ok=True)
        return str(path)
    
    def read_file(self, file_path: str) -> str:
        """Read the contents of a file.
        
        Args:
            file_path: Path to the file
            
        Returns:
            File contents
            
        Raises:
            FileNotFoundError: If the file doesn't exist
        """
        path = pathlib.Path(file_path)
        
        if not path.exists():
            self.logger.error(f"File not found: {file_path}")
            raise FileNotFoundError(f"File not found: {file_path}")
        
        try:
            return path.read_text(encoding="utf-8")
        except Exception as e:
            self.logger.error(f"Error reading file {file_path}: {str(e)}")
            raise